pkg/ring/ring.go

```diff
@@ -461,6 +462,100 @@ func (r *Ring) Subring(key uint32, n int) ReadRing {
 	return sub
 }
 
+// ShuffleShard returns a subring for the provided identifier (eg. a tenant ID)
+// and size (number of instances). The size is expected to be a multiple of the
+// number of zones and the returned subring will contain the same number of
+// instances per zone as far as there are enough registered instances in the ring.
+//
+// The algorithm used to build the subring is a shuffle sharder based on probabilistic
+// hashing. We treat each zone as a separate ring and pick N unique replicas from each
+// zone, walking the ring starting from random but predictable numbers. The random
+// generator is initialised with a seed based on the provided identifier.
+//
+// This implementation guarantees:
+// - Stability: given the same ring, two invocations returns the same result.
+// - Consistency: adding/removing 1 instance from the ring generates a resulting
+//   subring with no more then 1 difference.
+// - Shuffling: probabilistically, for a large enough cluster each identifier gets
+//   a different set of instances, with a reduced number of overlapping instances
+//   between two identifiers.
+func (r *Ring) ShuffleShard(identifier string, size int) ReadRing {
+	// Nothing to do if the shard size is not smaller then the actual ring.
+	if size <= 0 || r.IngesterCount() <= size {
+		return r
+	}
+
+	// Use the identifier to compute an hash we'll use to seed the random.
+	hasher := md5.New()
```

Comment:
We use 64-bit hashing elsewhere (fnv64a), perhaps we should use it here as well, and not throw away half of the hash.

Reply:
I tried using fnv64a:

```
	// Use the identifier to compute an hash we'll use to seed the random.
	seed := client.HashAdd64a(client.HashNew64a(), identifier)
	// Initialise the random generator used to select instances in the ring.
	random := rand.New(rand.NewSource(int64(seed)))
```

But the `TestRing_ShuffleShard_Shuffling` fails. Apparently, md5 introduces a better randomization of seeds. Given there's no performance penalty I revert it.